File: templates/permission-control.ts
import { query } from "@anthropic-ai/claude-agent-sdk";

/**
 * Permission Control Template
 *
 * Demonstrates:
 * - Permission modes: default, acceptEdits, bypassPermissions
 * - canUseTool(toolName, input, { signal }) → PermissionResult
 * - PermissionResult: { behavior: "allow", updatedInput } | { behavior: "deny", message }
 */

// Custom permission logic with canUseTool
async function withCustomPermissions() {
  for await (const message of query({
    prompt: "Deploy the application to production",
    options: {
      permissionMode: "default",

      canUseTool: async (toolName, input, { signal }) => {
        // Allow read-only tools
        if (["Read", "Grep", "Glob"].includes(toolName)) {
          return { behavior: "allow", updatedInput: input };
        }

        // Deny destructive bash commands
        if (toolName === "Bash") {
          const dangerous = ["rm -rf", "dd if=", "mkfs", "shutdown"];
          if (dangerous.some((p) => input.command?.includes(p))) {
            return { behavior: "deny", message: `Blocked: ${input.command}` };
          }
        }

        // Deny writes to sensitive paths
        if (toolName === "Write" || toolName === "Edit") {
          const sensitive = [".env", "credentials", "secrets"];
          if (sensitive.some((s) => input.file_path?.includes(s))) {
            return { behavior: "deny", message: `Sensitive file: ${input.file_path}` };
          }
        }

        return { behavior: "allow", updatedInput: input };
      },
    },
  })) {
    if (message.type === "result") {
      if (message.subtype === "success") {
        console.log(message.result);
      } else {
        console.error(message.subtype, message.errors);
      }
    }
  }
}

// Bypass all permissions (sandboxed environments only)
async function bypassed() {
  for await (const message of query({
    prompt: "Run tests and fix failures",
    options: {
      permissionMode: "bypassPermissions",
      allowDangerouslySkipPermissions: true,
    },
  })) {
    if (message.type === "result" && message.subtype === "success") {
      console.log(message.result);
    }
  }
}

withCustomPermissions().catch(console.error);

File: templates/python/permission_control.py
"""Custom permission control with Claude Agent SDK."""
import asyncio
from claude_agent_sdk import query, ClaudeAgentOptions, HookMatcher, HookContext

# NOTE: Use PreToolUse hooks for permission enforcement (recommended approach).
# can_use_tool callbacks require AsyncIterable prompts (raises ValueError with
# string prompts since v0.1.78+), and may not fire reliably — see KI #27.

async def permission_hook(
    input_data: dict, tool_use_id: str | None, context: HookContext
) -> dict:
    """Block dangerous operations using a PreToolUse hook."""
    tool_name = input_data.get("tool_name", "")
    tool_input = input_data.get("tool_input", {})

    # Allow read-only tools immediately
    if tool_name in ["Read", "Grep", "Glob"]:
        return {}

    # Block destructive bash commands
    if tool_name == "Bash":
        dangerous_patterns = ["rm -rf", "dd if=", "mkfs"]
        command = tool_input.get("command", "")
        if any(p in command for p in dangerous_patterns):
            return {
                "hookSpecificOutput": {
                    "hookEventName": "PreToolUse",
                    "permissionDecision": "deny",
                    "permissionDecisionReason": "Destructive command blocked",
                }
            }

    # Allow everything else
    return {}

async def main():
    from claude_agent_sdk import ResultMessage

    options = ClaudeAgentOptions(
        hooks={"PreToolUse": [HookMatcher(hooks=[permission_hook])]},
        permission_mode="default",
        max_turns=10,
    )

    async for msg in query(prompt="List files in the current directory", options=options):
        if isinstance(msg, ResultMessage) and msg.subtype == "success":
            print(msg.result)

asyncio.run(main())
